indexer/services/ender/src/lib/kafka-publisher.ts

```diff
@@ -43,34 +51,93 @@ export class KafkaPublisher {
   }
 
   public addEvent(event: ConsolidatedKafkaEvent) {
-    switch (event.topic) {
+    this.getMessages(event.topic)!.push(event.message);
+  }
+
+  /**
+   * Helper function to get messages for a given topic.
+   *
+   * @param kafkaTopic
+   * @private
+   */
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  private getMessages(kafkaTopic: KafkaTopics): any[] | undefined {
+    switch (kafkaTopic) {
       case KafkaTopics.TO_WEBSOCKETS_SUBACCOUNTS:
-        this.subaccountMessages.push(event.message);
-        break;
+        return this.subaccountMessages;
       case KafkaTopics.TO_WEBSOCKETS_TRADES:
-        this.tradeMessages.push(event.message);
-        break;
+        return this.tradeMessages;
       case KafkaTopics.TO_WEBSOCKETS_MARKETS:
-        this.marketMessages.push(event.message);
-        break;
+        return this.marketMessages;
       case KafkaTopics.TO_WEBSOCKETS_CANDLES:
-        this.candleMessages.push(event.message);
-        break;
+        return this.candleMessages;
       case KafkaTopics.TO_VULCAN:
-        this.vulcanMessages.push(event.message);
-        break;
+        return this.vulcanMessages;
       default:
         throw new Error('Invalid Topic');
     }
   }
 
   /**
-   * Sort trade events by block height, transaction index, and event index in ascending order,
-   * where the first trade event should be the earliest trade in the block.
+   * Sort subaccountMessages that represent fills by block height, transaction index,
+   * and event index in ascending order per order id. Only keep the subaccount message if
+   * it represents the last fill event per order id.
+   */
+  public retainLastFillEventsForSubaccountMessages() {
```

Comment:
This would lead to all prior fills for an order (if there were multiple) to also be omitted, and that would be incorrect.

E.g. Imagine order A has 2 fills:
- message 1: {fills: [ .... 40% of order size], order: [ .... status is OPEN, total filled is 40%]}
- message 2: {fills: [ .... 60% of order size], order: [..... status is FILLED, total filled is 100%]}

The websocket message I expect to see as a client is:
- message: {fills: [.... 40% of order size, .... 60% of order size], order: [....status is FILLED, total filled is 100%]}

However with this logic what the client sees is:
- message: {fills: [....60% of order size], order: [....status is FILLED, total filled is 100%]}

Reply:
done